grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java

```diff
@@ -136,6 +175,122 @@ public static GraqlTraversal createTraversal(PatternAdmin pattern, EmbeddedGrakn
+                    .forEach(fragment -> instanceVarTypeMap.put(fragment.end(), labelVarTypeMap.get(fragment.start())));
+        } while (size != instanceVarTypeMap.size());
+
+        // relationship vars and its role player vars
+        Multimap<Var, Var> relationshipRolePlayerMap = HashMultimap.create();
+        allFragments.stream().filter(OutRolePlayerFragment.class::isInstance)
+                .forEach(fragment -> relationshipRolePlayerMap.put(fragment.start(), fragment.end()));
+
+        // find all the relationships requiring type inference
+        Iterator<Var> iterator = relationshipRolePlayerMap.keySet().iterator();
+        while (iterator.hasNext()) {
+            Var relationship = iterator.next();
+
+            // the relation should have at least 2 known role players so we can infer something useful
+            if (instanceVarTypeMap.containsKey(relationship) ||
+                    relationshipRolePlayerMap.get(relationship).size() < 2) {
+                iterator.remove();
+            } else {
+                final int[] numRolePlayersHaveType = {0};
+                relationshipRolePlayerMap.get(relationship).forEach(
+                        rolePlayer -> {
+                            if (instanceVarTypeMap.containsKey(rolePlayer)) {
+                                numRolePlayersHaveType[0]++;
+                            }
+                        });
+                if (numRolePlayersHaveType[0] < 2) {
+                    iterator.remove();
+                }
+            }
+        }
+        if (relationshipRolePlayerMap.isEmpty()) return;
+
+        // for each type, get all possible relationship type it could be in
+        Map<Type, Set<RelationshipType>> relationshipMap = new HashMap<>();
+        labelVarTypeMap.values().stream().distinct().forEach(
+                type -> getAllPossibleRelationships(relationshipMap, type));
+
+        relationshipRolePlayerMap.asMap().forEach((relationshipVar, rolePlayerVars) -> {
+            Iterator<Var> rolePlayerVarIterator = rolePlayerVars.iterator();
+
+            Var firstRolePlayer = rolePlayerVarIterator.next();
+            while (!instanceVarTypeMap.containsKey(firstRolePlayer)) {
+                firstRolePlayer = rolePlayerVarIterator.next();
+            }
+
+            Set<Type> possibleRelationShipTypes = getAllPossibleRelationShipTypesOfARolePlayer(
+                    firstRolePlayer, instanceVarTypeMap, relationshipMap);
+            while (rolePlayerVarIterator.hasNext()) {
+                Var nextRolePlayer = rolePlayerVarIterator.next();
+                if (instanceVarTypeMap.containsKey(nextRolePlayer)) {
+
+                    // compute the intersection of possible relationship types
+                    possibleRelationShipTypes.retainAll(getAllPossibleRelationShipTypesOfARolePlayer(
+                            nextRolePlayer, instanceVarTypeMap, relationshipMap));
+                }
+            }
```

Comment:
Maybe you can use `Sets#intersection` and `reduce`:

```
Set<Type> possibleRelationshipTypes = rolePlayersVars.stream()
    .filter(instanceVarTypeMap::containsKey)
    .map(rolePlayer -> getAllPossibleRelationshipTypesOfARolePlayer(rolePlayer, instanceVarTypeMap, relationshipMap))
    .reduce(Sets::intersection);
```

Reply:
👍